# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [2]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [3]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [4]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [6]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [7]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [8]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [9]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [10]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [11]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [12]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [13]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [14]:
df.Season.nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [15]:
df.Wteam.value_counts().max

<bound method Series.max of Wteam
1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: count, Length: 364, dtype: int64>

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [16]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [17]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [18]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [19]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [20]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [21]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [22]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [23]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [24]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [25]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [26]:
import pandas as pd

# Prepare sorted DataFrames based on season, winning team, and winning score
df_sorted_season = df.sort_values('Season').head()
df_sorted_team = df.sort_values('Wteam').head()
df_sorted_score = df.sort_values('Wscore').head()

# Extract rows from index 100 to 200 and specific columns: season, winning team, and winning score using iloc
selected_data = df.iloc[100:201, [0, 5, 6]]  

# Showcase the sorted DataFrames based on season, winning team, and winning score
print("DataFrame sorted by Season:")
print(df_sorted_season)

print("DataFrame sorted by Winning Team:")
print(df_sorted_team)

print("DataFrame sorted by Winning Score:")
print(df_sorted_score)

# Show the selected data from rows 100 to 200 with specified columns
print("\nSelected rows from index 100 to 200 and columns - Season, Winning Team, Winning Score:")
print(selected_data)


DataFrame sorted by Season:
      Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0       1985      20   1228      81   1328      64    N      0
2484    1985      99   1275      73   1132      63    H      0
2485    1985      99   1306      62   1171      59    H      0
2486    1985      99   1318      99   1145      91    H      0
2487    1985      99   1337      75   1109      56    H      0
DataFrame sorted by Winning Team:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
140642    2016      25   1101      72   1197      62    N      0
136740    2015      62   1101      87   1146      70    H      0
143283    2016      89   1101      80   1249      71    H      0
143476    2016      92   1101      79   1223      72    H      0
141972    2016      61   1101      62   1146      61    A      0
DataFrame sorted by Winning Score:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
128582    2013     116   1264      34   1193      31    H      0
69336

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [27]:
# Sort the DataFrame in descending order based on the winning score
sorted_df_by_score = df.sort_values('Wscore', ascending=False)

# Selecting the top row which has the highest winning score
top_score_game = sorted_df_by_score.head(1).iloc[0]

# Finding the season and team with the highest winning score.
top_score_season = top_score_game['Season']
top_winning_team = top_score_game['Wteam']

# Showing the season and team for the highest-scoring game.
print("The season with the highest winning score is:", top_score_season)
print("The team with the highest winning score is:", top_winning_team)



The season with the highest winning score is: 1991
The team with the highest winning score is: 1258


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [28]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [29]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [30]:
import pandas as pd

# Add a new column 'ScoreDifference' that shows the absolute difference between winning and losing scores.
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Find the average 'ScoreDifference' for all games.
average_score_difference = df['ScoreDifference'].mean()

# Filter the DataFrame to show games with a 'ScoreDifference' higher than the average.
filtered_df = df[df['ScoreDifference'] > average_score_difference]

# Show the filtered DataFrame.
print(filtered_df)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
1

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [31]:
# Find the game with the highest 'ScoreDifference' in the filtered DataFrame.
game_with_highest_difference = filtered_df.loc[filtered_df['ScoreDifference'].idxmax()]

# Retrieve the season, winning team, and losing team details of the game with the highest 'ScoreDifference'.
highest_difference_season = game_with_highest_difference['Season']
winning_team_highest_difference = game_with_highest_difference['Wteam']
losing_team_highest_difference = game_with_highest_difference['Lteam']

# Show the season, winning team, and losing team details for the game with the highest 'ScoreDifference'.
print("Season of the game with the highest 'ScoreDifference':", highest_difference_season)
print("Winning team in the game with the highest 'ScoreDifference':", winning_team_highest_difference)
print("Losing team in the game with the highest 'ScoreDifference':", losing_team_highest_difference)


Season of the game with the highest 'ScoreDifference': 1996
Winning team in the game with the highest 'ScoreDifference': 1409
Losing team in the game with the highest 'ScoreDifference': 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [32]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [33]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       N         3
       A         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [34]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [35]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [36]:
# Group the DataFrame by season and calculate the average winning score for each season
average_winning_score_by_season = df.groupby('Season')['Wscore'].mean()

# Display the average winning score for each season
print(average_winning_score_by_season)


Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [37]:
# Group the DataFrame by winning team and find the maximum winning score for each team across all seasons
max_winning_score_by_team = df.groupby('Wteam')['Wscore'].max()

# Display the maximum winning score for each team across all seasons
print(max_winning_score_by_team)


Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [38]:
# Grouping our dataframe by both season and winning team, then looking to find the team with the highest average winning score for each of the season
team_highest_avg_score_by_season = df.groupby(['Season', 'Wteam'])['Wscore'].mean().reset_index()
team_highest_avg_score_by_season = team_highest_avg_score_by_season.loc[team_highest_avg_score_by_season.groupby('Season')['Wscore'].idxmax()]

# Displaying and print of the team with the highest average winning score for each season
print(team_highest_avg_score_by_season)


      Season  Wteam      Wscore
174     1985   1328   92.800000
287     1986   1109   91.200000
783     1987   1380   95.875000
977     1988   1258  111.750000
1268    1989   1258  117.315789
1560    1990   1258  126.347826
1950    1991   1380  112.312500
2246    1992   1380   99.642857
2542    1993   1380  101.875000
2842    1994   1380  106.583333
3003    1995   1206  102.833333
3305    1996   1206  103.600000
3648    1997   1254   94.142857
4070    1998   1395  101.153846
4312    1999   1317   95.750000
4695    2000   1395   95.647059
5013    2001   1395   99.315789
5201    2002   1242   92.888889
5660    2003   1395   88.222222
5809    2004   1194   92.142857
6239    2005   1311   91.333333
6460    2006   1190   89.533333
7026    2007   1440   99.400000
7367    2008   1440   94.500000
7652    2009   1377   96.200000
8061    2010   1440   95.000000
8414    2011   1449   90.130435
8683    2012   1370   87.400000
9036    2013   1377   86.500000
9333    2014   1322   92.600000
9683    

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [39]:
# Grouping the DataFrame by both season and winning team and counting the number of wins for each team in every season
team_season_wins = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Displaying the new dataframe showing the number of wins for each team in each respective season
print(team_season_wins)


       Season  Wteam  Wins
0        1985   1102     5
1        1985   1103     9
2        1985   1104    21
3        1985   1106    10
4        1985   1108    19
...       ...    ...   ...
10167    2016   1460    20
10168    2016   1461    12
10169    2016   1462    27
10170    2016   1463    21
10171    2016   1464     9

[10172 rows x 3 columns]


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [40]:
# Grouping our dataframe by both season and winning team and then counting the of wins each team in each season
team_season_wins = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Finding the team with the highest amount of wins for each season
team_with_most_wins_per_season = team_season_wins.loc[team_season_wins.groupby('Season')['Wins'].idxmax()]

# Displaying and printing the team with the most wins for each season
print(team_with_most_wins_per_season)


      Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30
2088    1992   1181    28
2423    1993   1231    28
2665    1994   1163    27
2928    1995   1116    26
3357    1996   1269    31
3638    1997   1242    31
3943    1998   1242    33
4201    1999   1181    32
4709    2000   1409    29
4829    2001   1181    29
5121    2002   1153    30
5456    2003   1166    29
5981    2004   1390    29
6165    2005   1228    32
6452    2006   1181    30
6841    2007   1242    30
7209    2008   1272    33
7552    2009   1272    31
7870    2010   1242    32
8217    2011   1242    32
8565    2012   1246    32
8878    2013   1211    30
9462    2014   1455    33
9609    2015   1246    34
9955    2016   1242    29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [41]:
# Generating mean (average) scores for both the losing and winning teams
losing_team_means = df.groupby('Lteam')['Lscore'].mean()
winning_team_means = df.groupby('Wteam')['Wscore'].mean()

# Creating a DataFrame and grouping it by average losing and winning scores
comparison_data = pd.DataFrame({'Average Losing Score': losing_team_means, 'Average Winning Score': winning_team_means})

# Checking if teams have a higher average losing score than winning score.
comparison_data['Higher Losing Score'] = comparison_data['Average Losing Score'] > comparison_data['Average Winning Score']

# Displaying the comparison results of mean losing and winning scores of the different teams
print("Analyzing Average Scores for Winning and Losing Teams:")
print(comparison_data)





Analyzing Average Scores for Winning and Losing Teams:
      Average Losing Score  Average Winning Score  Higher Losing Score
1101             60.586207              78.111111                False
1102             59.201507              69.893204                False
1103             64.117347              75.839768                False
1104             64.374317              75.825944                False
1105             61.675373              74.960894                False
...                    ...                    ...                  ...
1460             63.082915              75.531469                False
1461             64.661972              75.170082                False
1462             68.216117              79.906021                False
1463             60.829213              71.720102                False
1464             65.187063              73.926056                False

[364 rows x 3 columns]


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [42]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [43]:
# Function is to determine if a game is considered high scoring or not
def is_high_scoring_game(row):
    if row['Wscore'] > 100:
        return 'Yes'
    else:
        return 'No'

# Applying the following function so we can add the 'HighScoringGame' column to the DataFrame
df['HighScoringGame'] = df.apply(is_high_scoring_game, axis=1)

# Showing a view of the DataFrame with the new column.

print(df.head())


   Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   
3    1985      25   1165      70   1432      54    H      0               16   
4    1985      25   1192      86   1447      74    H      0               12   

  HighScoringGame  
0              No  
1              No  
2              No  
3              No  
4              No  


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [44]:
# Creating a dictionary for keeping track of the number of games played by each team
games_played = {}

# Iterating over the rows of the current DataFrame and making and update the total number of games for each team
for index, row in df.iterrows():
    if row['Wteam'] in games_played:
        games_played[row['Wteam']] += 1
    else:
        games_played[row['Wteam']] = 1
    
    if row['Lteam'] in games_played:
        games_played[row['Lteam']] += 1
    else:
        games_played[row['Lteam']] = 1

# Showing and displaying the total number of games played by each team
for team, total_games in games_played.items():
    print(f"Team {team} played a total of {total_games} games.")


Team 1228 played a total of 992 games.
Team 1328 played a total of 968 games.
Team 1106 played a total of 855 games.
Team 1354 played a total of 906 games.
Team 1112 played a total of 981 games.
Team 1223 played a total of 363 games.
Team 1165 played a total of 833 games.
Team 1432 played a total of 69 games.
Team 1192 played a total of 908 games.
Team 1447 played a total of 903 games.
Team 1218 played a total of 931 games.
Team 1337 played a total of 922 games.
Team 1226 played a total of 847 games.
Team 1242 played a total of 993 games.
Team 1268 played a total of 969 games.
Team 1260 played a total of 914 games.
Team 1133 played a total of 949 games.
Team 1305 played a total of 922 games.
Team 1424 played a total of 974 games.
Team 1307 played a total of 969 games.
Team 1288 played a total of 925 games.
Team 1344 played a total of 951 games.
Team 1438 played a total of 952 games.
Team 1374 played a total of 916 games.
Team 1411 played a total of 903 games.
Team 1412 played a total o

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [45]:
# Initializing a dictionary for storage the highest score difference per season
highest_score_diff_per_season = {}

# Iterating over the rows in the DataFrame in order to find the game with the highest score difference for each season
for season in df['Season'].unique():
    season_df = df[df['Season'] == season]
    max_score_diff = 0
    max_score_diff_values = None
    
    for index, row in season_df.iterrows():
        score_diff = row['Wscore'] - row['Lscore']
        if score_diff > max_score_diff:
            max_score_diff = score_diff
            max_score_diff_values = {'Winning Team': row['Wteam'], 'Losing Team': row['Lteam'], 'Score Difference': score_diff}
    
    if max_score_diff_values is not None:
        highest_score_diff_per_season[season] = max_score_diff_values

# Showing the game with the highest score difference for each season underneath
for season, game_info in highest_score_diff_per_season.items():
    print(f"For season {season}, the game with the highest score difference is between teams {game_info['Winning Team']} and {game_info['Losing Team']} with a difference of {game_info['Score Difference']} points.")



For season 1985, the game with the highest score difference is between teams 1361 and 1288 with a difference of 60 points.
For season 1986, the game with the highest score difference is between teams 1314 and 1264 with a difference of 84 points.
For season 1987, the game with the highest score difference is between teams 1155 and 1118 with a difference of 73 points.
For season 1988, the game with the highest score difference is between teams 1328 and 1147 with a difference of 68 points.
For season 1989, the game with the highest score difference is between teams 1242 and 1135 with a difference of 70 points.
For season 1990, the game with the highest score difference is between teams 1181 and 1217 with a difference of 76 points.
For season 1991, the game with the highest score difference is between teams 1163 and 1148 with a difference of 68 points.
For season 1992, the game with the highest score difference is between teams 1116 and 1126 with a difference of 82 points.
For season 1993,

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [46]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [47]:
import pandas as pd

# Finding the total games played by each team, including wins and losses.


winning_team_games = df['Wteam'].value_counts()
losing_team_games = df['Lteam'].value_counts()

# Adding the number of games played by each team together and looking at the score
total_games_played = winning_team_games.add(losing_team_games, fill_value=0).astype(int)

# Converting the Series to a DataFrame for better visualization and easier display of it
team_games_df = total_games_played.reset_index()
team_games_df.columns = ['Team', 'TotalGames']

# printing the dataframe with the total number of games played by each team
print("Total number of games played by each team:")
print(team_games_df)




Total number of games played by each team:
     Team  TotalGames
0    1101          76
1    1102         840
2    1103         910
3    1104         975
4    1105         447
..    ...         ...
359  1460         827
360  1461         914
361  1462         954
362  1463         838
363  1464         856

[364 rows x 2 columns]


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [48]:
import pandas as pd

# Creating a new column 'ScoreDifference' by performing vectorized subtraction and using the groupby function.


df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Identifying the game with the highest score difference for each season using groupby and idxmax functions.


idx = df.groupby('Season')['ScoreDifference'].idxmax()

# Obtaining the rows corresponding to these indices.
games_with_highest_difference_per_season = df.loc[idx]

# Result is shown underneath
print("Game with the highest score difference for each season:")
print(games_with_highest_difference_per_season)

Game with the highest score difference for each season:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H   

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [49]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [50]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [51]:
type(df['Wscore'])

pandas.core.series.Series

In [52]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [53]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [54]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [55]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y